File: simple/2.两数之和/add_sum.go
package main

import "fmt"

type ListNode struct {
	Val int
	Next *ListNode
}


func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var point = &ListNode{Val: 0, Next: nil}
	head := point
	carry := false
	for l1 != nil || l2 != nil || carry {
		sum := 0
		l1Val := 0
		l2Val := 0
		if l1 != nil {
			l1Val  = l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			l2Val = l2.Val
			l2 = l2.Next
		}
		if carry {
			sum += 1
			carry = false
		}
		sum += l1Val + l2Val
		if sum > 9 {
			sum = sum - 10
			carry = true
		}
		point.Next = &ListNode{Val: sum, Next: nil}
		point = point.Next
	}
	return head.Next
}

func main() {
	l1 := &ListNode{
		Val:  2,
		Next: &ListNode{
			Val: 4,
			Next: &ListNode{
				Val: 3,
				Next: nil,
			},
		},
	}
	l2 := &ListNode{
		Val:  5,
		Next: &ListNode{
			Val: 6,
			Next: &ListNode{
				Val: 4,
				Next: nil,
			},
		},
	}
	fmt.Println(addTwoNumbers(l1, l2))
}

File: simple/2/add_sum.py
# -*- coding: utf-8 -*-
# @File  : add_sum.py
# @Author: mengde
# @Date  : 2021/11/11

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
            res = point = ListNode()
            carry = False
            while True:
                new_node = ListNode()
                sum = 0
                if l1 or l2 or carry:
                    l1_num = l1.val if l1 else 0
                    l2_num = l2.val if l2 else 0
                    l1 = l1.next if l1 else None
                    l2 = l2.next if l2 else None
                    if carry:
                        sum +=  1
                        carry = False
                    sum +=l1_num + l2_num
                    if sum < 10:
                        new_node.val = sum
                    else:
                        new_node.val = sum - 10
                        carry = True
                    point.next  = new_node
                    point = point.next
                else:
                    break
            return res.next
